specs/_features/eip6110/beacon-chain.md

```diff
@@ -0,0 +1,413 @@
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+    )
+```
+
+#### New `apply_deposit`
+
+```python
+def apply_deposit(state: BeaconState,
+                  pubkey: BLSPubkey,
+                  withdrawal_credentials: Bytes32,
+                  amount: uint64,
+                  signature: BLSSignature,
+                  ) -> None:
+    validator_pubkeys = [validator.pubkey for validator in state.validators]
+    if pubkey not in validator_pubkeys:
+        # Verify the deposit signature (proof of possession) which is not checked by the deposit contract
+        deposit_message = DepositMessage(
+            pubkey=pubkey,
+            withdrawal_credentials=withdrawal_credentials,
+            amount=amount,
+        )
+        domain = compute_domain(DOMAIN_DEPOSIT)  # Fork-agnostic domain since deposits are valid across forks
+        signing_root = compute_signing_root(deposit_message, domain)
+        # Initialize validator if the deposit signature is valid
+        if bls.Verify(pubkey, signing_root, signature):
+            state.validators.append(get_validator_from_deposit_data(pubkey, withdrawal_credentials, amount))
+            state.balances.append(amount)
+            state.previous_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.current_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.inactivity_scores.append(uint64(0))
+    else:
+        # Increase balance by deposit amount
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        increase_balance(state, index, amount)
+```
+
+
+#### New `process_deposit_receipt`
+
+```python
+def process_deposit_receipt(state: BeaconState, deposit_receipt: DepositReceipt) -> None:
+    # Set deposit receipt start index
+    if state.deposit_receipts_start_index == NOT_SET_DEPOSIT_RECEIPTS_START_INDEX:
+        state.deposit_receipts_start_index = deposit_receipt.index
+
+    # Signify the end of transition to in-protocol deposit logic
+    if state.eth1_deposit_index >= state.deposit_receipts_start_index:
+        state.eth1_deposit_index = deposit_receipt.index + 1
+
+    apply_deposit(
+        state=state, 
+        pubkey=deposit_receipt.pubkey,
+        withdrawal_credentials=deposit_receipt.withdrawal_credentials,
+        amount=deposit_receipt.amount,
+        signature=deposit_receipt.signature,
+    )
+```
+
+#### Modified `process_deposit`
+
+*Note*: The function `process_deposit` is modified to prevent deposits from being processed the second time (due to `process_deposit_receipt`).
+
+```python
+def process_deposit(state: BeaconState, deposit: Deposit) -> None:
+    # [New in EIP-6110]
+    # Skip already processed deposits
+    if state.eth1_deposit_index >= state.deposit_receipts_start_index:
+        state.eth1_deposit_index += 1
```

Comment:
I'm confused about how this works `deposit_receipts_start_index` is static once initialized. So why do we need to do this increment here? can't it just be a `return` and forever return

Reply:
Once `state.eth1_deposit_index` reaches the first in-protocol deposit index there is likely a tail of deposits from the same voting period remaining to be processed. Say, no in-protocol deposit is happening for hours from that event. In such a case `Eth1Data` poll machinery would be attempting to include the same portion of deposits because the index wouldn't be update.

If we make a change to the validator spec to stop including old fashioned deposits once the first in-protocol index is reached then we can do a simple return. Current spec is made in a way that no changes to the validator logic is required, and all the new changes are introduced into the state transition function where we can ensure to have a good test coverage without additional developer's work.

Reply:
If we do change the validator spec then we can leave `process_deposit` unchanged by enforcing the following condition in the `process_operations`:
```python
# Disable former deposit mechanism once all prior deposits are processed
eth1_deposit_index_limit = min(state.eth1_data.deposit_count, state.deposit_receipts_start_index)
if eth1_deposit_index_limit < state.eth1_deposit_index:
    assert len(body.deposits) == min(MAX_DEPOSITS, eth1_deposit_index_limit - state.eth1_deposit_index)
else:
    assert len(body.deposits) == 0
```

This actually would look cleaner from the spec perspective

Reply:
Made it into the spec. Lmk, if it looks fine and I will update validator guide respectively

Reply:
UPD: changes to the validator guide are in place now